cluster/rpc.go

```diff
@@ -3,50 +3,197 @@ package cluster
-	*rpc.Client
-	registry Registry
+	conns *connectionBalancer
 }
 
-func NewClient(serviceName string, r Registry) (*Client, error) {
-	node, err := nodeToDial(serviceName, r)
+func newClient(host string, serviceName string, r Registry) (*Client, error) {
+	conns, err := newConnectionBalancer(host, serviceName, r)
 	if err != nil {
 		return nil, err
 	}
+	return &Client{
+		conns: conns,
+	}, nil
+}
 
-	dialAddr := fmt.Sprintf("%v:%v", node.Address, node.Port)
-	client, err := rpc.DialHTTP("tcp", dialAddr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to dial service address %v: %w", dialAddr, err)
+func (c *Client) Call(serviceMethod string, args interface{}, reply interface{}) error {
+	client := c.conns.Get()
+	return client.Call(serviceMethod, args, reply)
+}
+
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	client := c.conns.Get()
+	return client.Go(serviceMethod, args, reply, done)
+}
+
+func (c *Client) Close() error {
+	return c.conns.Close()
+}
+
+func (c *Client) ConnectionErrs() chan error {
+	return c.conns.errChan
+}
+
+const defaultMaxConnections = 3
+const defaultInitalNodeTimeout = 5 * time.Second
+
+type connectionBalancer struct {
+	clientNode string
+	seq        uint64
+
+	selectedNodes []Node
+	clients       []*rpc.Client
+	lock          sync.RWMutex
+
+	cancel    func()
+	errChan   chan error
+	waitGroup sync.WaitGroup
+}
+
+func newConnectionBalancer(host string, serviceName string, r Registry) (*connectionBalancer, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &connectionBalancer{
+		clientNode: host,
+		clients:    []*rpc.Client{},
+		cancel:     cancel,
+		errChan:    make(chan error, 1),
 	}
 
-	return &Client{
-		Client:   client,
-		registry: r,
-	}, nil
+	nodesChan := r.WatchService(ctx, serviceName)
+
+	var initalNodes []Node
+	select {
+	case initalNodes = <-nodesChan:
+	case <-time.After(defaultInitalNodeTimeout):
+		return nil, fmt.Errorf("no inital nodes provided for %v", serviceName)
+	}
+
+	if err := c.handleNewNodes(initalNodes); err != nil {
+		return nil, err
+	}
+
+	c.waitGroup.Add(1)
+	go c.watchForNewNodes(ctx, nodesChan)
+	return c, nil
 }
 
-func nodeToDial(serviceName string, r Registry) (node Node, err error) {
-	serv, err := r.Services(context.Background())
+func (c *connectionBalancer) Get() *rpc.Client {
+	return c.roundRobinSelect()
+}
+
+func (c *connectionBalancer) roundRobinSelect() *rpc.Client {
+	index := atomic.AddUint64(&c.seq, uint64(1))
+	clients := c.getClients()
+	return clients[int(index)%len(clients)]
+}
+
+func (c *connectionBalancer) Close() error {
+	if c.cancel != nil {
+		c.cancel()
+	}
+	for _, client := range c.clients {
+		client.Close()
+	}
+	c.waitGroup.Wait()
+	close(c.errChan)
+	return nil
+}
+
+func (c *connectionBalancer) watchForNewNodes(ctx context.Context, nodesChan chan []Node) {
+	defer c.waitGroup.Done()
+
+	for {
+		select {
+		case nodes := <-nodesChan:
```

Comment:
I wonder if we should consider "debouncing" this. If multiple nodes get added to the cluster at the same time (which is likely) we should just `handleNewNodes` once.